streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java

```diff
@@ -156,9 +157,10 @@ public long offset() {
     @Override
     public Headers headers() {
         if (recordContext == null) {
-            throw new IllegalStateException("This should not happen as headers() should only be called while a record is processed");
+            return StreamsHeaders.emptyHeaders();
```

Comment:
I did some hacking around in this PR to make things work reasonably. I'm not sure I always made exactly the right call, but the main point was to prototype the proposed public API, which seems to work nicely.